drivers/i3c/i3c_dw.c

```diff
@@ -0,0 +1,2349 @@
+	sys_write32(config->regs + SLV_INTR_REQ, slv_intr_req);
+
+	/* wait for SLV_IBI_RESP update */
+	status = k_sem_take(&data->ibi_sts_sem, K_MSEC(100));
+	if (status != 0) {
+		return -ETIMEDOUT;
+	}
+
+	slv_ibi_resp = sys_read32(config->regs + SLV_INTR_REQ);
+	switch (SLV_IBI_RESP_IBI_STS(slv_ibi_resp)) {
+	case SLV_IBI_RESP_IBI_STS_ACK:
+		LOG_DBG("%s: Controller ACKed IBI TIR", dev->name);
+		return 0;
+	case SLV_IBI_RESP_IBI_STS_NACK:
+		LOG_ERR("%s: Controller NACKed IBI TIR", dev->name);
+		return -EAGAIN;
+	case SLV_IBI_RESP_IBI_STS_EARLY_TERMINATE:
+		LOG_ERR("%s: Controller aborted IBI TIR with %lu remaining", dev->name,
+			SLV_IBI_RESP_DATA_LENGTH(slv_ibi_resp));
+		return -EIO;
+	default:
+		return -EIO;
+	}
+}
+
+static int dw_i3c_target_ibi_raise(const struct device *dev, struct i3c_ibi *request)
+{
+	if (request == NULL) {
+		return -EINVAL;
+	}
+
+	switch (request->ibi_type) {
+	case I3C_IBI_TARGET_INTR:
+		return dw_i3c_target_ibi_raise_tir(dev, request);
+	case I3C_IBI_CONTROLLER_ROLE_REQUEST:
+		/* TODO: Synopsys I3C can support CR, but not implemented yet */
+		return -ENOTSUP;
+	case I3C_IBI_HOTJOIN:
+		return dw_i3c_target_ibi_raise_hj(dev);
+	default:
+		return -EINVAL;
+	}
+}
+
+#endif /* CONFIG_I3C_USE_IBI */
+
+static int i3c_dw_irq(const struct device *dev)
+{
+	const struct dw_i3c_config *config = dev->config;
+	struct dw_i3c_data *data = dev->data;
+	uint32_t status;
+	uint32_t present_state;
+
+	status = sys_read32(config->regs + INTR_STATUS);
+	if (status & (INTR_TRANSFER_ERR_STAT | INTR_RESP_READY_STAT)) {
+		dw_i3c_end_xfer(dev);
+
+		if (status & INTR_TRANSFER_ERR_STAT) {
+			sys_write32(INTR_TRANSFER_ERR_STAT, config->regs + INTR_STATUS);
+		}
+	}
+
+	if (status & INTR_IBI_THLD_STAT) {
+#ifdef CONFIG_I3C_USE_IBI
+		ibis_handle(dev);
+#endif /* CONFIG_I3C_USE_IBI */
+	}
+
+	/* target mode related interrupts */
+	present_state = sys_read32(config->regs + PRESENT_STATE);
+	if (!(present_state & PRESENT_STATE_CURRENT_MASTER)) {
+		const struct i3c_target_callbacks *target_cb = data->target_config->callbacks;
+
+		/* Data has been written to the rx fifo by a controller while in target mode*/
+		if (status & INTR_RX_THLD_STAT) {
+			/* while the rx fifo buffer is not empty */
+			while (DATA_BUFFER_STATUS_LEVEL_RX(
+				       sys_read32(config->regs + DATA_BUFFER_STATUS_LEVEL)) > 0) {
+				uint8_t rx_data;
+				/*
+				 * TODO: verify how byte size (lol) data is written to a dword size
+				 * buffer over i3c in target mode
+				 */
+				read_rx_fifo(dev, &rx_data, sizeof(rx_data));
+				/* call function pointer for write */
+				if (target_cb != NULL && target_cb->write_received_cb != NULL) {
+					target_cb->write_received_cb(data->target_config, rx_data);
+				}
+			}
+			sys_write32(INTR_RX_THLD_STAT, config->regs + INTR_STATUS);
+		}
+		/* Read Requested when the CMDQ is empty*/
+		if (status & INTR_READ_REQ_RECV_STAT) {
+			if (target_cb != NULL && target_cb->write_received_cb != NULL) {
```

Comment:
Should it not be "target_cb->read_requested_cb != NULL" ?

Reply:
no, you just check if target_cb is not NULL before you even look at target_cb->read_requested_cb, otherwise you are dereferencing a NULL will do a hard crash if you do

Reply:
/* Read Requested when the CMDQ is empty*/
        if (status & INTR_READ_REQ_RECV_STAT) {
            if (target_cb != NULL && **target_cb->write_received_cb** != NULL) {
                /* TODO: how should the first byte be handled? */
                **target_cb->read_requested_cb**(data->target_config, NULL);
            }
            sys_write32(INTR_READ_REQ_RECV_STAT, config->regs + INTR_STATUS);
        }

Here, validity of "write_received_cb" is checked but "read_requested_cb" is processed, which is wrong. I was pointing this.

Reply:
ok, I see now, yes that is correct. thanks for pointing that out